### PR TITLE
WOEM 14391 | USPS Tracking V3

### PR DESCRIPTION
**Pull request by ksolkowski**
Kind of like how shopify takes in session. This does it too

https://developer.usps.com/trackingv3#tag/Resources/operation/get-package-tracking

Removed all the "create return label" logic. So now all the env vars that matter for USPS is just the referrer ip.

Can't decide if i like throwing it in options or if i should give it named kwargs `.track(number, bearer_token:, options:` 50/50.

[WOEM-14391 (3.0)](https://wantable.atlassian.net/browse/WOEM-14391) - USPS API v3 Upgrade:
> 
> !https://mail.google.com/mail/u/0?ui=2&ik=9d935a2b01&attid=0.1.0.1&permmsgid=msg-f:1813554883647496906&th=192b0a8edc7be2ca&view=fimg&fur=ip&sz=s0-l75-ft&attbid=ANGjdJ_y65g3YELkE4RqbvT7XRA0G-xwQH4_hcfDta8INJzY4DEXqlzu0Jnnblsv7CsuNLDYsO2rlOvtLbb1cAJUqJsVASq1r81zh3BSIgZ--jvccvl4B2xGpylFI1I&disp=emb|width=1431,height=206!
> *October 16, 2024*
> *Retirement of APIs Version 1 and Version 2*
> Effective January 22, 2025, the USPS API Versions 1 and 2 will officially retire.  To ensure uninterrupted access to USPS services and take advantage of our latest innovations, we highly recommend upgrading to USPS API Version 3 at your earliest convenience.
> *Why Upgrade to USPS API Version 3?*
> *Version 3* introduces significant enhancements over previous versions, offering robust performance and advanced capabilities tailored to modern business needs:
> * *Improved Reliability and Uptime*: With enhanced fault tolerance, Version 3 ensures better system reliability, minimizing downtime and disruptions for your operations.
> * *Optimized API Performance*: Faster response times and improved data throughput mean that your systems will experience reduced latency when making requests, providing a smoother user experience.
> * *Expanded Functionality*: Version 3 includes new endpoints and features designed for increased flexibility and precision, such as:
> ** *Enhanced Tracking Capabilities*: More detailed tracking data, with granular updates and predictive delivery times.
> ** *Expanded Address Validation*: Improved algorithms for parsing and validating international addresses.
> ** *Advanced Rate Calculations*: Enhanced pricing logic for real-time shipping cost estimation, including multi-package shipments.
> * *Greater Security*: Leveraging updated encryption protocols and OAuth 2.0, Version 3 strengthens the protection of API keys and sensitive data.
> *Technical Considerations for Transition*
> Migrating to *API Version 3* may require adjustments to your existing integration. Here are a few key points to consider for a smooth transition:
> * *Backward Incompatibility*: API Version 3 introduces changes to endpoint structures and request/response formats. We recommend reviewing the updated API documentation to identify any necessary modifications in your implementation.
> * *OAuth 2.0 Implementation*: Version 3 uses OAuth 2.0 for API authentication, replacing legacy authentication methods. This will require the generation and management of new tokens for secure access.
> * *New Error Handling Standards*: More granular error codes and response details will be available, helping your system more effectively diagnose and respond to issues during API interactions.
> * *Sandbox Environment*: To support your development efforts, USPS offers a *sandbox environment* where you can thoroughly test the integration with Version 3 before deploying to production.
> *Support and Resources*
> We recognize that this transition is critical for your business and customers, and we are committed to assisting you every step of the way:
> * *API Documentation*: Comprehensive guides and migration instructions are available on the [USPS Developers Portal|https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdevelopers.usps.com%2F&data=05%7C02%7CAisha.C.Arthur%40usps.gov%7Ca9035ea8617d4e21c49608dcee235e2f%7Cf9aa5788eb334a498ad076101910cac3%7C0%7C0%7C638647082908612968%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=5U34muE6hc4F9cCWwJmIrQ1APSsE06f8mH%2FhJjJ1TZg%3D&reserved=0].
> Contact Information:
> * For questions or support, reach out to our technical team at: [apisupport@usps.gov|mailto:apisupport@usps.gov]
> * For information on USPS APIs visit: [https://developers.usps.com/|https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdevelopers.usps.com%2F&data=05%7C02%7CAisha.C.Arthur%40usps.gov%7Ca9035ea8617d4e21c49608dcee235e2f%7Cf9aa5788eb334a498ad076101910cac3%7C0%7C0%7C638647082908633973%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=qpCdWfhEshWZC%2FC8NmjAJLmzT6I765eu1rtRciOhCW4%3D&reserved=0]
> * For information specific to USPS API onboarding, features, etc., please log in to your USPS API account on the developer portal ([https://developers.usps.com/|https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdevelopers.usps.com%2F&data=05%7C02%7CAisha.C.Arthur%40usps.gov%7Ca9035ea8617d4e21c49608dcee235e2f%7Cf9aa5788eb334a498ad076101910cac3%7C0%7C0%7C638647082908647303%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=dDZqC%2BsDaLnRUA35YQLpkfMLNDBJD00As3f4DCvltfA%3D&reserved=0]) and navigate to "Tech Sheets" in the menu.
> We will provide you with monthly reminders/updates until APIs Version 1 and Version 2 retire on January 22, 2025.

### CODE REVIEW

File: lib/omniship/usps.rb [hunk 1]
```diff
@@ -8,65 +7,37 @@ module USPS
...
     end
 
     def self.tracking_url(number)
       TRACKING_URL + number
     end
 
-    # <EventTime>7:09 am</EventTime>
-    # <EventDate>January 17, 2017</EventDate>
+    def self.bearer_token_required?
+      true
```
Comment (nbucciarelli) [inline lib/omniship/usps.rb:31]: 👍

**Review by nbucciarelli — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -182,45 +181,6 @@ trk.shipment.packages.first.tracking_number
 # => "9400111201080302430600"
 ```
 
-Make a return shipping label
```
Comment (fiestacasey) [inline README.md:185]: fair enough removal

File: lib/omniship/dhl.rb [hunk 3]
```diff
@@ -13,5 +13,9 @@ def self.tracking_test?(tracking)
     def self.tracking_url(number)
       TRACKING_URL + number
     end
+
+    def self.bearer_token_required?
```
Comment (fiestacasey) [inline lib/omniship/dhl.rb:21]: could have created a base class for this method and inherited from it in each of the providers

**Review by fiestacasey — APPROVED** [review]
seems good to me. there's a bunch of different ways to pass things down and share logic/provide a uniform interface in a repo like this

**Comment by fiestacasey** [comment]
Is this not on Travis?

**Comment by ksolkowski** [comment]
@fiestacasey yeahh ain't on travis. And tests should be reworked with webmock lol. 

The next step I think is moving all bearer logic out of omniship to match what i did with usps

**Pull request merged**